tests/unit/json_test.c

```diff
@@ -1908,116 +1909,92 @@ static void test_json_null_not_null(void)
-{
-    const size_t len = JsonLength(a);
-    if (len != JsonLength(b))
-    {
-        return false;
-    }
-
-    for (size_t i = 0; i < len; i++)
-    {
-        const JsonElement *const a_child = JsonArrayGet(a, i);
-        const JsonElement *const b_child = JsonArrayGet(b, i);
-        if (!compare_json_object_merge_deep(a_child, b_child)) {
-            return false;
-        }
-    }
-    return true;
-}
-
-static bool compare_object_json_object_merge_deep(const JsonElement *const a, const JsonElement *const b)
-{
-    if (JsonLength(a) != JsonLength(b))
-    {
-        return false;
-    }
-
-    JsonIterator iter = JsonIteratorInit(a);
-    while (JsonIteratorHasMore(&iter))
-    {
-        const char *const key = JsonIteratorNextKey(&iter);
-        const JsonElement *const a_child = JsonObjectGet(a, key);
-        const JsonElement *const b_child = JsonObjectGet(b, key);
-        if (b_child == NULL || !compare_json_object_merge_deep(a_child, b_child))
-        {
-            return false;
-        }
-    }
-    return true;
-}
-
-static bool compare_json_object_merge_deep(const JsonElement *const a, const JsonElement *const b)
-{
-    const JsonType type = JsonGetType(a);
-    if (type != JsonGetType(b))
-    {
-        return false;
-    }
-
-    switch (type)
-    {
-    case JSON_TYPE_OBJECT:
-        return compare_object_json_object_merge_deep(a, b);
-    case JSON_TYPE_ARRAY:
-        return compare_array_json_object_merge_deep(a, b);
-    default:
-        return JsonCompare(a, b) == 0;
-    }
-}
-
 static bool check_json_object_merge_deep(const char *base_raw, const char *extra_raw, const char *expected_raw)
 {
+    LogSetGlobalLevel(LOG_LEVEL_DEBUG);
+    bool pass = true;
+    JsonElement *actual = NULL;
     JsonElement *base = NULL;
+    JsonElement *extra = NULL;
+    JsonElement *expected = NULL;
+    char *expected_string = NULL;
+    char *actual_string = NULL;
+
     if (JsonParse(&base_raw, &base) != JSON_PARSE_OK)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *extra = NULL;
     if (JsonParse(&extra_raw, &extra) != JSON_PARSE_OK)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *expected = NULL;
     if (JsonParse(&expected_raw, &expected) != JSON_PARSE_OK) {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *actual = JsonObjectMergeDeep(base, extra);
+    expected_string = JsonToString(expected);
+
+    actual = JsonObjectMergeDeep(base, extra);
     if (actual == NULL || actual == base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    /* JsonCompare is naive and does not handle containers properly, thus we
-     * need a custom comparison function to test for equality. */
-    if (!compare_json_object_merge_deep(actual, expected))
+    actual_string = JsonToString(actual);
+
+    if (JsonCompare(actual, expected))
```

Comment:
Use explicit comparison
```suggestion
    if (JsonCompare(actual, expected) != 0)
```